ls1intum/Artemis
# [Exam] Generate student exams

### Checklist

### Motivation and Context
We need a first basic implementation of the generate student exams feature.

### Description
- [x] We've implemented a first basic version of the generate student exams algorithm
- [x] We've added a corresponding Spring integration test for the generation of student exams
- [x] We've implemented the algorithm in a way that it ensures that `numberOfExercisesInExam` is equal to the actual number of exercises added to all generated student exams
- [x] We've implemented the client functionality to trigger the generation
- [x] We've implemented the error handling in the client

On re-generation of student exams all existing student exams of the exam should be deleted. This could be done by making use of `orphanRemoval`. As this leads to `A collection with cascade="all-delete-orphan" was no longer referenced by the owning entity instance` I've used `deleteInBatch()`. It ensures that the student exams of the exam and the relation with the exercises is removed from the database.

### Steps for Testing
You can test it on the test server and locally. To test the generation algorithm completely, you have to test locally and inspect the following database tables: `student_exam`, `student_exam_exercise`.

1. Log in to Artemis
2. Go to the exam overview of any course (e.g. https://artemistest.ase.in.tum.de/#/course-management/33/exams)
3. Create a new exam (we have some backward compatibility problems caused by other features)
4. Create multiple mandatory and optional exercise groups (see possible configuration below)
5. Register at least on user for the exam
6. Click on `Student exams` in the overview
7. Click on `Generate student exams`
8. If an error message appears solve it by adjusting the exam configuration. If no error message appears, check that you don't get an error on console and that student exams got created
9. Click again on `Generate student exams`
10. The existing student exams should have been deleted and new ones should have been created (compare by ID)

**A valid configuration would be:**
* `numberOfExercisesInExam` = 5
* number of mandatory exercise groups: 3
* number of optional exercise groups: 5 (--> then one exercise from two of the optional exercise groups will be selected)

### Screenshots
![image](https://user-images.githubusercontent.com/15168372/85062269-7e4fc300-b1a8-11ea-808d-d3295ccc880e.png)

![image](https://user-images.githubusercontent.com/15168372/85062847-85c39c00-b1a9-11ea-834e-b53d75ad6b37.png)
